chapter_recurrent-modern/machine-translation-and-dataset.ipynb

```diff
@@ -0,0 +1,628 @@
+    "    for (int i = 0; i < linesIntArr.size(); i++) {\n",
+    "        ArrayList<Integer> temp = new ArrayList<>();\n",
+    "        temp.addAll(Arrays.asList(linesIntArr.get(i)));\n",
+    "        temp.add(vocab.getIdx(\"<eos>\"));\n",
+    "        linesIntArr.set(i, temp.stream().toArray(n -> new Integer[n]));\n",
+    "    }\n",
+    "\n",
+    "    NDArray arr = manager.create(new Shape(linesIntArr.size(), numSteps), DataType.INT32);\n",
+    "    int row = 0;\n",
+    "    for (Integer[] line : linesIntArr) {\n",
+    "        NDArray rowArr = manager.create(truncatePad(line, numSteps, vocab.getIdx(\"<pad>\")));\n",
+    "        arr.set(new NDIndex(\"{}:\", row), rowArr);\n",
+    "        row += 1;\n",
+    "    }\n",
+    "    NDArray validLen = arr.neq(vocab.getIdx(\"<pad>\")).sum(new int[] {1});\n",
+    "    return new Pair<>(arr, validLen);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 18
+   },
+   "source": [
+    "## Putting All Things Together\n",
+    "\n",
+    "Finally, we define the `loadDataNMT` function\n",
+    "to return the data iterator, together with\n",
+    "the vocabularies for both the source language and the target language.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Pair<ArrayDataset, Pair<Vocab, Vocab>> loadDataNMT(\n",
+    "        int batchSize, int numSteps, int numExamples) throws IOException {\n",
+    "    /* Return the iterator and the vocabularies of the translation dataset. */\n",
+    "    StringBuilder text = preprocessNMT(readDataNMT().toString());\n",
+    "    Pair<ArrayList<String[]>, ArrayList<String[]>> pair =\n",
+    "            tokenizeNMT(text.toString(), numExamples);\n",
+    "    ArrayList<String[]> source = pair.getKey();\n",
+    "    ArrayList<String[]> target = pair.getValue();\n",
+    "    Vocab srcVocab =\n",
+    "            new Vocab(\n",
+    "                    source.stream().toArray(String[][]::new),\n",
+    "                    2,\n",
+    "                    new String[] {\"<pad>\", \"<bos>\", \"<eos>\"});\n",
+    "    Vocab tgtVocab =\n",
+    "            new Vocab(\n",
+    "                    target.stream().toArray(String[][]::new),\n",
+    "                    2,\n",
+    "                    new String[] {\"<pad>\", \"<bos>\", \"<eos>\"});\n",
+    "\n",
+    "    Pair<NDArray, NDArray> pairArr = buildArrayNMT(source, srcVocab, numSteps);\n",
+    "    NDArray srcArr = pairArr.getKey();\n",
+    "    NDArray srcValidLen = pairArr.getValue();\n",
+    "\n",
+    "    pairArr = buildArrayNMT(target, tgtVocab, numSteps);\n",
+    "    NDArray tgtArr = pairArr.getKey();\n",
+    "    NDArray tgtValidLen = pairArr.getValue();\n",
+    "\n",
+    "    ArrayDataset dataset =\n",
+    "            new ArrayDataset.Builder()\n",
+    "                    .setData(srcArr, srcValidLen, tgtArr, tgtValidLen)\n",
```

Comment:
tgt sounds like a label, you should put it in .optLabels(tgt)

Reply:
Done